src/rgw/rgw_gc.cc

```diff
@@ -65,7 +65,65 @@ int RGWGC::tag_index(const string& tag)
   return rgw_shards_mod(XXH64(tag.c_str(), tag.size(), seed), max_objs);
 }
 
-int RGWGC::send_chain(cls_rgw_obj_chain& chain, const string& tag)
+int RGWGC::send_split_chain(const cls_rgw_obj_chain& chain, const std::string& tag)
+{
+  constexpr int FACTOR_INCREASE = 3; //approximate value by which total size of chain, tag and time increases after encoding into bufferlist
+  ldpp_dout(this, 20) << "RGWGC::send_split_chain - tag is: " << tag << dendl;
+
+  if (cct->_conf->rgw_max_chunk_size) {
+    cls_rgw_obj_chain broken_chain;
+    bool chain_sent = false;
+    auto tag_size = tag.length(), expiration_time_size = sizeof(ceph::real_time);
+    auto adjusted_max_write_size = ((cct->_conf->rgw_max_chunk_size)/FACTOR_INCREASE); //approx upper limit
+    ldpp_dout(this, 20) << "RGWGC::send_split_chain - rgw_max_chunk_size is: " << cct->_conf->rgw_max_chunk_size << dendl;
+    ldpp_dout(this, 20) << "RGWGC::send_split_chain - adjusted_max_write_size is: " << adjusted_max_write_size << dendl;
+    auto cumulative_size = 0;
+    for (auto it = chain.objs.begin(); it != chain.objs.end(); it++) {
+      auto size = it->pool.size() + it->loc.size() + it->key.name.size() + it->key.instance.size();
```

Comment:
think it's worth being more systematic about this part? if we account for everything, we should be able to avoid the `FACTOR_INCREASE` part entirely:
```c++
struct cls_rgw_obj_key {
...
  size_t estimate_encoded_size() const {
    constexpr size_t start_overhead = sizeof(__u8) + sizeof(__u8) + sizeof(ceph_le32); // ENCODE_START()
    constexpr size_t string_overhead = sizeof(__u32); // strings are encoded with 32-bit length prefix
    return start_overhead +
        string_overhead + name.size() +
        string_overhead + instance.size();
  }
};

struct cls_rgw_obj {
...
  size_t estimate_encoded_size() const {
    constexpr size_t start_overhead = sizeof(__u8) + sizeof(__u8) + sizeof(ceph_le32); // ENCODE_START()
    constexpr size_t string_overhead = sizeof(__u32); // strings are encoded with 32-bit length prefix
    return start_overhead +
        string_overhead + pool.size() +
        string_overhead + key.name.size() +
        string_overhead + loc.size() +
        key.estimate_encoded_size();
  }
};
```

Reply:
doesn't FACTOR_INCREASE exist to correct for the unexpected space amp after encoding these objects of known sizes?

Reply:
there will be some extra encoding overhead from MOSDOp, cls_rgw_gc_set_entry_op, cls_rgw_gc_obj_info, and cls_rgw_obj_chain. but outside of `cls_rgw_obj`, none of that overhead is per-object. so instead of `rgw_max_chunk_size/FACTOR_INCREASE`, we might do something like `rgw_max_chunk_size - OVERHEAD` to account for it

however, our object writes that send `rgw_max_chunk_size` of object data don't account for any of that MOSDOp overhead, so it may not be neccessary here either

Reply:
is that true though?  if so, then yes

Reply:
> think it's worth being more systematic about this part? if we account for everything, we should be able to avoid the `FACTOR_INCREASE` part entirely:
> 
> ```c++
> struct cls_rgw_obj_key {
> ...
>   size_t estimate_encoded_size() const {
>     constexpr size_t start_overhead = sizeof(__u8) + sizeof(__u8) + sizeof(ceph_le32); // ENCODE_START()
>     constexpr size_t string_overhead = sizeof(__u32); // strings are encoded with 32-bit length prefix
>     return start_overhead +
>         string_overhead + name.size() +
>         string_overhead + instance.size();
>   }
> };
> 
> struct cls_rgw_obj {
> ...
>   size_t estimate_encoded_size() const {
>     constexpr size_t start_overhead = sizeof(__u8) + sizeof(__u8) + sizeof(ceph_le32); // ENCODE_START()
>     constexpr size_t string_overhead = sizeof(__u32); // strings are encoded with 32-bit length prefix
>     return start_overhead +
>         string_overhead + pool.size() +
>         string_overhead + key.name.size() +
>         string_overhead + loc.size() +
>         key.estimate_encoded_size();
>   }
> };
> ```

Does this come from the way encoding happens?

Reply:
> there will be some extra encoding overhead from MOSDOp, cls_rgw_gc_set_entry_op, cls_rgw_gc_obj_info, and cls_rgw_obj_chain. but outside of `cls_rgw_obj`, none of that overhead is per-object. so instead of `rgw_max_chunk_size/FACTOR_INCREASE`, we might do something like `rgw_max_chunk_size - OVERHEAD` to account for it
> 
> however, our object writes that send `rgw_max_chunk_size` of object data don't account for any of that MOSDOp overhead, so it may not be neccessary here either

I have observed that 'do_op msg data len' that comes from PrimaryLogPG.cc, always has 16 bytes extra. Is that what MOSDOp overhead is?

Reply:
ok

Reply:
In order to arrive at the correct size that can be sent over to the OSD, can we not calculate the encoding overhead for the entire chain beforehand using similar logic that you have stated above and then once we have the estimate = rgw_max_chunk_size - encoding_overhead and then use that to split the chain into smaller parts.

Reply:
or calculate the estimated encoding size per object and then compare the same against rgw_max_chunk_size for splitting the chain (in this case we dont need to subtract overhead from rgw_max_chunk_size)

Reply:
> or calculate the estimated encoding size per object and then compare the same against rgw_max_chunk_size for splitting the chain (in this case we dont need to subtract overhead from rgw_max_chunk_size)

that sounds right, yeah

Reply:
@cbodley : I have taken into account the overhead of all encoding that happens before the op is sent to the osd. And the calculation matches the size after actual encoding. But when the op reaches the osd, I see a difference of 100 bytes, I logged the size in PrimaryLogPG::do_op(). This will only be a problem when rgw_max_chunk_size is the same as osd_max_write_size and when the encoded length of the op will be the same as rgw_max_chunk_size. Should I spend more time understanding the osd part to take into account 100 bytes, or should I just take into account 100 extra bytes.